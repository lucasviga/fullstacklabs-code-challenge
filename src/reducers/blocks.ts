export const blocksInitialState = {
  loading: false,
  error: false,
  data: [
    {
      id: "",
      attributes: {
        data: ""
      }
    }
  ]
}
