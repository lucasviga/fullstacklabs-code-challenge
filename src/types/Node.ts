import { Blocks } from "./Blocks";

export interface Node {
  online: boolean;
  name: string;
  url: string;
  loading: boolean;
  blocks: Blocks;
}
