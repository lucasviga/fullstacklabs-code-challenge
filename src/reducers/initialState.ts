import { blocksInitialState } from "./blocks";

const initialState = () => ({
  nodes: {
    list: [
      {
        url: "https://thawing-springs-53971.herokuapp.com",
        online: false,
        name: "Node 1",
        loading: false,
        blocks: blocksInitialState,
      },
      {
        url: "https://secret-lowlands-62331.herokuapp.com",
        online: false,
        name: "Node 2",
        loading: false,
        blocks: blocksInitialState,
      },
      {
        url: "https://calm-anchorage-82141.herokuapp.com",
        online: false,
        name: "Node 3",
        loading: false,
        blocks: blocksInitialState,
      },
      {
        url: "http://localhost:3002",
        online: false,
        name: "Node 4",
        loading: false,
        blocks: blocksInitialState,
      },
    ],
  },
});
export default initialState;
