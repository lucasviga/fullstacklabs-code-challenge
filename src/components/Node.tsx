import React from "react";
import ExpandMoreIcon from "@mui/icons-material/ExpandMore";
import {
  Accordion,
  AccordionSummary,
  Typography,
  AccordionDetails,
  Box,
} from "@mui/material";
import { styled } from "@mui/material/styles";
import colors from "../constants/colors";
import Status from "./Status";
import { Node as NodeType } from "../types/Node";
import Blocks from "./Blocks";

type Props = {
  node: NodeType;
  expanded: boolean;
  toggleNodeExpanded: (node: NodeType) => void;
};

const AccordionRoot = styled(Accordion)({
  margin: "16px 0",
  boxShadow: "0px 3px 6px 1px rgba(0,0,0,0.15)",

  "&:before": {
    backgroundColor: "unset",
  },
});

const AccordionSummaryContainer = styled(AccordionSummary)({
  padding: "0 24px",
  "& .MuiAccordionSummary-content": {
    margin: "10px 0 !important", // Avoid change of sizing on expanded
  },
  "& .MuiAccordionSummary-expandIconWrapper": {
    color: colors.faded,
  },
});

const BoxSummaryContent = styled(Box)({
  display: "flex",
  flexDirection: "row",
  justifyContent: "space-between",
  alignItems: "center",
  width: "100%",
  paddingRight: 20,
});

const TypographyHeading = styled(Typography)({
  fontSize: 17,
  display: "block",
  color: colors.text,
  lineHeight: 1.5,
});

const TypographySecondaryHeading = styled(Typography)(({ theme }) => ({
  fontSize: 14,
  color: colors.faded,
  lineHeight: 2,
}));

const Node: React.FC<Props> = ({ node, expanded, toggleNodeExpanded }) => {
  return (
    <AccordionRoot
      elevation={3}
      expanded={expanded}
      onChange={() => toggleNodeExpanded(node)}
    >
      <AccordionSummaryContainer expandIcon={<ExpandMoreIcon />}>
        <BoxSummaryContent>
          <Box>
            <TypographyHeading variant="h5">
              {node.name || "Unknown"}
            </TypographyHeading>
            <TypographySecondaryHeading variant="subtitle1">
              {node.url}
            </TypographySecondaryHeading>
          </Box>
          <Status loading={node.loading} online={node.online} />
        </BoxSummaryContent>
      </AccordionSummaryContainer>
      <AccordionDetails>
        <Blocks blocks={node.blocks} />
      </AccordionDetails>
    </AccordionRoot>
  );
};

export default Node;
