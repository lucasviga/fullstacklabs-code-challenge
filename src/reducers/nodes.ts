import { createAsyncThunk, createSlice } from "@reduxjs/toolkit";
import initialState from "./initialState";
import { Node } from "../types/Node";
import { RootState } from "../store/configureStore";
import fetch from "cross-fetch";

export interface NodesState {
  list: Node[];
}

export const checkNodeStatus = createAsyncThunk(
  "nodes/checkNodeStatus",
  async (node: Node) => {
    const response = await fetch(`${node.url}/api/v1/status`);
    const responseBlocks = await fetch(`${node.url}/api/v1/blocks`);

    const blocks = await responseBlocks.json();
    const {node_name} = await response.json();

    const data = { node_name, blocks }

    return data;
  }
);

export const checkNodesStatus = createAsyncThunk(
  "nodes/checkNodesStatus",
  async (nodes: Node[], thunkAPI) => {
    const { dispatch } = thunkAPI;
    nodes.forEach((node) => {
      dispatch(checkNodeStatus(node));
    });
  }
);

export const nodesSlice = createSlice({
  name: "nodes",
  initialState: initialState().nodes as NodesState,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(checkNodeStatus.pending, (state, action) => {
      const node = state.list.find((node) => node.url === action.meta.arg.url);
      if (node) {
        node.loading = true;

        node.blocks.loading = true;
      }
    });
    builder.addCase(checkNodeStatus.fulfilled, (state, action) => {
      const node = state.list.find((node) => node.url === action.meta.arg.url);
      if (node) {
        node.online = true;
        node.loading = false;
        node.name = action.payload.node_name;

        node.blocks.loading = false;
        node.blocks.error = false;
        node.blocks.data = action.payload.blocks.data;
      }
    });
    builder.addCase(checkNodeStatus.rejected, (state, action) => {
      const node = state.list.find((node) => node.url === action.meta.arg.url);
      if (node) {
        node.online = false;
        node.loading = false;

        node.blocks.error = true;
        node.blocks.loading = false;
        node.blocks.data = [];
      }
    });
  },
});

export const selectNodes = (state: RootState) => state.nodes.list;
export default nodesSlice.reducer;
