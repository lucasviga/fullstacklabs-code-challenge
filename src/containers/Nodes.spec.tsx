import React from "react";
import { mount } from "enzyme";
import configureMockStore, { MockStoreEnhanced } from "redux-mock-store";
import thunk from "redux-thunk";
import { Provider } from "react-redux";
import { create } from "react-test-renderer";
import ConnectedNodes from "./Nodes";
import Node from "../components/Node";
import { checkNodesStatus } from "../reducers/nodes";

describe("<Nodes />", () => {
  const nodes = {
    list: [
      {
        url: "https://thawing-springs-53971.herokuapp.com",
        online: false,
        name: "Node 1",
        loading: false,
        blocks: {
          loading: false,
          error: false,
          data: [
            {
              id: "",
              attributes: {
                data: ""
              }
            }
          ]
        }
      },
      {
        url: "https://secret-lowlands-62331.herokuapp.com",
        online: false,
        name: "Node 2",
        loading: false,
        blocks: {
          loading: false,
          error: false,
          data: [
            {
              id: "",
              attributes: {
                data: ""
              }
            }
          ]
        }
      },
    ],
  };

  let store: MockStoreEnhanced<unknown, {}>;

  function setup(): JSX.Element {
    const middlewares = [thunk];
    store = configureMockStore(middlewares)({ nodes });
    return (
      <Provider store={store}>
        <ConnectedNodes />
      </Provider>
    );
  }

  afterEach(() => {
    store.clearActions();
  });

  it("should contain <Node />", () => {
    const wrapper = mount(setup());

    expect(wrapper.find(Node).length).toEqual(2);
    expect(store.getActions()).toEqual(
      expect.arrayContaining([
        expect.objectContaining({
          meta: expect.objectContaining({ arg: nodes.list }),
          type: checkNodesStatus.pending.type,
        }),
      ])
    );
  });

  it("should match snapshot", () => {
    const component = create(setup());
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });
});
