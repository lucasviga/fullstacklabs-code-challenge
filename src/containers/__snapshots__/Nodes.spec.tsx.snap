// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nodes /> should match snapshot 1`] = `
<div
  className="MuiBox-root css-1ak52z1"
>
  <h1
    className="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
  >
    <strong
      style={
        Object {
          "color": "#000",
        }
      }
    >
      Nodes
    </strong>
  </h1>
  <div
    className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation3 MuiAccordion-root MuiAccordion-rounded MuiAccordion-gutters css-185lh1p-MuiPaper-root-MuiAccordion-root"
  >
    <div
      aria-expanded={false}
      className="MuiButtonBase-root MuiAccordionSummary-root MuiAccordionSummary-gutters css-n997py-MuiButtonBase-root-MuiAccordionSummary-root"
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="button"
      tabIndex={0}
    >
      <div
        className="MuiAccordionSummary-content MuiAccordionSummary-contentGutters css-o4b71y-MuiAccordionSummary-content"
      >
        <div
          className="MuiBox-root css-16rygv8"
        >
          <div
            className="MuiBox-root css-0"
          >
            <h5
              className="MuiTypography-root MuiTypography-h5 css-bl5h9z-MuiTypography-root"
            >
              Node 1
            </h5>
            <h6
              className="MuiTypography-root MuiTypography-subtitle1 css-gvet60-MuiTypography-root"
            >
              https://thawing-springs-53971.herokuapp.com
            </h6>
          </div>
          <div
            className="MuiBox-root css-70qvj9"
          >
            <span
              className="css-1y3s8m3"
            />
            <span
              className="css-10g39r7"
            >
              OFFLINE
            </span>
          </div>
        </div>
      </div>
      <div
        className="MuiAccordionSummary-expandIconWrapper css-yw020d-MuiAccordionSummary-expandIconWrapper"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ExpandMoreIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
          />
        </svg>
      </div>
    </div>
    <div
      className="MuiCollapse-root MuiCollapse-vertical MuiCollapse-hidden css-bz4dnt-MuiCollapse-root"
      style={
        Object {
          "minHeight": "0px",
        }
      }
    >
      <div
        className="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
      >
        <div
          className="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
        >
          <div
            className="MuiAccordion-region"
            role="region"
          >
            <div
              className="MuiAccordionDetails-root css-15v22id-MuiAccordionDetails-root"
            >
              <p
                className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
              >
                Blocks go here
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation3 MuiAccordion-root MuiAccordion-rounded MuiAccordion-gutters css-185lh1p-MuiPaper-root-MuiAccordion-root"
  >
    <div
      aria-expanded={false}
      className="MuiButtonBase-root MuiAccordionSummary-root MuiAccordionSummary-gutters css-n997py-MuiButtonBase-root-MuiAccordionSummary-root"
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="button"
      tabIndex={0}
    >
      <div
        className="MuiAccordionSummary-content MuiAccordionSummary-contentGutters css-o4b71y-MuiAccordionSummary-content"
      >
        <div
          className="MuiBox-root css-16rygv8"
        >
          <div
            className="MuiBox-root css-0"
          >
            <h5
              className="MuiTypography-root MuiTypography-h5 css-bl5h9z-MuiTypography-root"
            >
              Node 2
            </h5>
            <h6
              className="MuiTypography-root MuiTypography-subtitle1 css-gvet60-MuiTypography-root"
            >
              https://secret-lowlands-62331.herokuapp.com
            </h6>
          </div>
          <div
            className="MuiBox-root css-70qvj9"
          >
            <span
              className="css-1y3s8m3"
            />
            <span
              className="css-10g39r7"
            >
              OFFLINE
            </span>
          </div>
        </div>
      </div>
      <div
        className="MuiAccordionSummary-expandIconWrapper css-yw020d-MuiAccordionSummary-expandIconWrapper"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ExpandMoreIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
          />
        </svg>
      </div>
    </div>
    <div
      className="MuiCollapse-root MuiCollapse-vertical MuiCollapse-hidden css-bz4dnt-MuiCollapse-root"
      style={
        Object {
          "minHeight": "0px",
        }
      }
    >
      <div
        className="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
      >
        <div
          className="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
        >
          <div
            className="MuiAccordion-region"
            role="region"
          >
            <div
              className="MuiAccordionDetails-root css-15v22id-MuiAccordionDetails-root"
            >
              <p
                className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
              >
                Blocks go here
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
